### ISSUE TITLE
Loading of Singularity Issues and Potential Network Congestion Issues?

### ISSUE THREAD

**Issue by frenchwr** [apptainer/singularity#627]
Hi there! We have started using Singularity at Vanderbilt University's HPC center (Advanced Computing Center for Research and Education, or ACCRE; http://www.accre.vanderbilt.edu/) and we LOVE it! Really great work with this tool.

One issue that is a little unclear to me...

We have a lot of relatively small, embarrassingly parallel type analyses on our cluster. It is not uncommon at all for users to submit and run 1000 or more single core jobs through SLURM at once with job arrays, with each job running for a relatively short amount of time. Should we be concerned about potential network congestion issues if an image is loaded off of a shared filesystem (we use GPFS) potentially hundreds or thousands of times in a relatively short period of time? Some of the images being developed by our users are on the order of 5GB. Does the entire image get moved from GPFS to a compute node each time a container of that image is run?

Thanks again, and all the best!

**Comment by damonms** [comment]
Hi All,

I work in the lab that is stemming this issue. I'm happy to provide test cases if desired.

Cheers,
Steve

**Comment by gmkurtzer** [comment]
Hello,

As with any job programs and data being run at scale, there is the possibility for congestion of shared infrastructure. Now, Singularity was architected for parallel execution in mind which is one reason why we use a single image file by default. This single image file contains the root file system of your container but the entire image does *NOT* get moved from GPFS or any POSIX compatible shared storage to each compute node that is running it. But, there is of course some data that will always be pulled as the image is loaded. Initially the image and file system meta-data and header, then once it is loaded, only the bits required for the container to operate will be transferred. No wasted (or rather very little) wasted network Tx/Rx.

Additionally, for very large jobs, that require large amounts of metadata lookups, Singularity will also be very beneficial there as the image itself is only a single metadata lookup to your shared storage (GPFS, Lustre, NFS, etc..). Then the image's metadata is cached, and thus metadata calls on inodes within the image are all done locally, which can yield very high performance benefits when comparing to shared storage.

To summarize; yes, there will be an increase in network load due to the containers being loaded but the architecture itself was designed to be as optimal as possible to work with shared storage.

Let me know if that answers your question and/or causes more.

**Comment by frenchwr** [comment]
@gmkurtzer Thank you for the detailed response. It sounds like Singularity was designed with these types of use cases in mind. @damonms I'll let you follow up with any specific questions in case you have them.

**Comment by damonms** [comment]
Hi @gmkurtzer,

Just following up on this. We ran a test of a single singularity image ~32gb across GPFS. First we ran 50 at once, had no issues, all containers essentially started immediately. Then we scaled up to approximately 200 containers all starting simultaneously and we saw about a 7 minute lag for the final container to actually start the execution algorithm contained inside it. Finally we ran approximately 430 containers, again all starting simultaneously and noticed about an hour of lag for the final container to being the execution algorithm. 

We also ran a test where we copied the container directly to the node, and unsurprisingly there was no lag. However, this is not an ideal solution for us as our HPC is obviously a shared resource and we would like to avoid caching multiple containers on all nodes for our use. 

I should also mention that when we ran the tests, the nodes were 100% reserved for our usage as to ensure we do not impact other jobs on the machines. 

@frenchwr Can give you some technical details of the backend networking system if that would be helpful.

Curious to hear your thoughts!

Thanks again for your help, we're really looking forward to using Singularity exclusively for our image processing containorization and deployment.

Cheers,
Steve

**Comment by frenchwr** [comment]
@gmkurtzer et al., any thoughts?

**Comment by bauerm97** [comment]
Can we investigate and see if this is still a problem occurring in the 3.x series?

**Comment by dtrudg** [comment]
Hi all. We've recently heard reports of launches across large sets of nodes without impact vs native execution of the software so I'm going to close this. If you can replicate in the latest version of Singularity, please do re-open the issue. Many Thanks.
